Track order counts by payment method

menakas opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Currently, the tracker data does not provide any insight into the number of orders by gateway. This is a frequently requested information. To help track this data, I added a function `get_order_counts_by_payment_method` with a query that gets this data. The results of this function is merged with the other `orders` related data. To be able to eventually sift out this information from other orders related data, I have used a prefix `gateway_` so that the tracker meta_key will be prefixed `orders_gateway_`.


### How to test the changes in this Pull Request:
1. Ensure your test site is opted into tracking (WooCommerce > Settings > Advanced > WooCommerce.com > Usage Tracking).
2. Using `wp shell` or vardump, you can call the new method directly to check the results:

`WC_Tracker:: get_order_counts_by_payment_method();`

Results will be similar to this:

         Array
        (
           [gateway_stripe] => 705
           [gateway_ppec_paypal] => 211
           [gateway_square_credit_card] => 41
           [gateway_paypal_pro] => 35
           [gateway_kco] => 24
           [gateway_paypal_advanced] => 5
           [gateway_cod] => 4
           [gateway_paypal-pro-hosted] => 1
           [gateway_bacs] => 1
       )

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

Documentation to be done as per p2-p7Ldg5-JS

### Changelog entry

> Added orders count by payment method to Tracker data and replaced direct DB calls with CRUD.

menakas committed:
Add a function to get order counts by payment method

deltaWhiskey commented on includes/class-wc-tracker.php:
```diff
@@ -540,6 +541,47 @@ private static function get_admin_user_agents() {
+				    ) AS payment_method
+				FROM {$wpdb->prefix}posts AS orders
+				LEFT JOIN {$wpdb->prefix}postmeta AS order_meta ON order_meta.post_id = orders.ID
+				WHERE orders.post_status in ( 'wc-completed', 'wc-processing', 'wc-refunded' )
```
I think we should add `orders.post_type = 'shop_order'` to the WHERE clause, so that we only get one `wp_post` record per refunded order. Also, that lets the query leverage the `type_status_date` key.

menakas replied:
Thanks @deltaWhiskey 
I've added a commit for that.

deltaWhiskey reviewed (commented):
I like this approach. I think the query should be adjusted.

menakas committed:
Check if post_type is shop_order

deltaWhiskey commented on includes/class-wc-tracker.php:
```diff
@@ -540,6 +541,49 @@ private static function get_admin_user_agents() {
+		);
+
+		foreach ( $orders_by_payment_method as $orders_count ) {
+			$method                                  = 'gateway_' . $orders_count->payment_method;
+			$order_counts_by_payment_method[ $method ] = $orders_count->orders_count;
```
Trivial formatting comment: the equals signs don't quite line up.

menakas replied:
Thanks. I missed it. 🤦‍♀️

peterfabian commented on includes/class-wc-tracker.php:
```diff
@@ -540,6 +541,49 @@ private static function get_admin_user_agents() {
+			GROUP BY payment_method
+			ORDER BY orders_count DESC
+		"
+		);
```
I'm sorry if it's apparent, but is there a reason to group the inner query by order id and picking the `MAX` from the group? 

```
SELECT
			    order_gateway.payment_method AS payment_method,
			    COUNT( id ) AS orders_count
			FROM (
				SELECT
					orders.ID AS id,
				    order_meta.meta_value AS payment_method
				FROM wp_posts AS orders
				LEFT JOIN wp_postmeta AS order_meta ON order_meta.post_id = orders.ID
				WHERE
					orders.post_type = 'shop_order' AND
					orders.post_status in ( 'wc-completed', 'wc-processing', 'wc-refunded' )
                AND meta_key = '_payment_method'
				AND meta_value IS NOT NULL
			) AS order_gateway
			
			GROUP BY payment_method
			ORDER BY orders_count DESC
```

It seems to me the query above should return the same set of results and should be slightly less costly, especially since it can ignore all those NULL values that would be otherwise created for all meta values. 

Is it because we want to be absolutely positive each order is counted only once or some other reason?

Conceptually, I don't like that this is a direct SQL query bypassing our CRUD abstraction, but since other tracker data is fetched in the same way, it's probably okay to leave it as it is, as it's more performant this way. We'll need to clean it up early next year anyway.

menakas replied:
You are right. I was trying to get the `payment method title` as well and eventually dropped the idea, as the title can be changed for each site depending on the admin's fancy. But the structure of the query has remained. I shall make the appropriate changes.

peterfabian reviewed (commented):
Overall, this should work fine, I have a question regarding the privacy---have you folks checked that it's okay to add this type of info without extra consent from merchants?

menakas commented:
@peterfabian 
Thanks for the review.
1. Regarding consent from merchants for adding this info, I shall check and get back.

2. As per my response inline, I shall change the query accordingly.

3.

>We'll need to clean it up early next year anyway.

Early next year is not far away. I don't mind going through the CRUD abstraction, if it is not too much of a detour. Let me know your thoughts on this.

menakas committed:
Use wc_get_orders() to add order dates for tracker

menakas committed:
Add order counts by status

menakas committed:
Add order counts by gateway

menakas committed:
Get order total values

menakas commented:
@peterfabian 
On second thoughts, I've added commits to use the CRUD abstraction in the Tracker instead of direct queries. Let me know your thoughts.

menakas committed:
Some fixes for coding standards

peterfabian commented on includes/class-wc-tracker.php:
```diff
@@ -366,16 +366,96 @@ private static function get_order_counts() {
+		$orders = wc_get_orders(
+			array(
+				'type'           => array( 'shop_order', 'shop_order_refund' ),
+				'customer'       => '',
```
I think the `customer` parameter is not needed.

peterfabian commented on includes/class-wc-tracker.php:
```diff
@@ -366,16 +366,96 @@ private static function get_order_counts() {
+			array(
+				'type'           => array( 'shop_order', 'shop_order_refund' ),
+				'customer'       => '',
+				'posts_per_page' => -1,
```
I'm slightly worried here in terms of memory usage on large shops. Have you tried to run this on a store with tens of thousands of orders? It might be better to paginate & iterate over it, what do you think?

menakas replied:
Ah yes, I haven't tested it in large stores. What is the standard page size that is used in WC? I'll use the same.

peterfabian commented on includes/class-wc-tracker.php:
```diff
@@ -540,6 +620,49 @@ private static function get_admin_user_agents() {
+	 *
+	 * @return array
+	 */
+	public static function get_order_counts_by_payment_method() {
```
I think this method can be removed now, right, since [the code above](https://github.com/woocommerce/woocommerce/pull/28584/files#diff-81c4e5a312109472fa8526612aeb12807759e0ee28d912a7dc775c890a909d16R417-R428) calculates the same thing?

menakas replied:
Yes, as I mentioned [here](https://github.com/woocommerce/woocommerce/pull/28584#issuecomment-749867388), what would you prefer that I do with the order such functions that this code replaces?

peterfabian replied:
Ah, I missed that part, sorry. Since they are public, we can't remove them. I think deprecation is probably the right way to go. Thanks for switching to CRUD!

peterfabian reviewed (changes requested):
Hi, thanks for making the changes, looks good so far, and sorry for the delay here. 
I have a couple more questions, can you please have a look at those?

menakas committed:
Remove the functio to get orders count by payment method

menakas committed:
Paginate while getting orders to calculate tracker data

menakas committed:
Use site's posts_per_page size

menakas committed:
Deprecate public function and remove private functions

menakas commented:
I've used a page size of 100. If you prefer the site's page size to be used, I've added a commit for it. Keep it or drop it as you think appropriate.

As to deprecation, the `get_orders()` function is private.

Of the three functions that were called inside this earlier, `get_order_dates()` and `get_order_counts()` are private as well. So, I removed them.

But `get_order_totals()` is public. I've deprecated it without any replacement.

menakas commented:
@peterfabian
Is this fine or do you need any changes?

peterfabian committed:
Fixed the deprecation version

peterfabian reviewed (approved):
Looks good to me, thanks for making those changes!

Just want to note that the data type will change for a bunch of values from string to int/float, so hope that is fine for the tracker on the receiving end.

Before:
<img width="748" alt="Screenshot 2021-01-19 at 11 26 53" src="https://user-images.githubusercontent.com/2207451/105022822-7a3f9880-5a4a-11eb-86ab-0d9947d450ff.png">


After:
<img width="748" alt="Screenshot 2021-01-19 at 11 24 41" src="https://user-images.githubusercontent.com/2207451/105022792-714ec700-5a4a-11eb-86d4-0029085459e9.png">
